Add docs for Python object -> DuckDB conversion (#1462)
duckdb/duckdb-web: docs/api/python/data_ingestion.md

This is a mapping of Python object types to DuckDB [Logical Types](../../sql/data_types/overview.html):

* `None` -> `NULL`
* `bool` -> `BOOLEAN`
* `datetime.timedelta` -> `INTERVAL`
* `str` -> `VARCHAR`
* `bytearray` -> `BLOB`
* `memoryview` -> `BLOB`
* `decimal.Decimal` -> `DECIMAL` / `DOUBLE`
* `uuid.UUID` -> `UUID`

The rest of the conversion rules are as follows.

### `int`

Since integers can be of arbitrary size in Python, there is not a one-to-one conversion possible for ints.
Intead we perform these casts in order until one succeeds:

* `BIGINT`
* `INTEGER`
* `UBIGINT`
* `UINTEGER`
* `DOUBLE`

When using the DuckDB Value class, it's possible to set a target type, which will influence the conversion.

### `float`

These casts are tried in order until one succeeds:

* `DOUBLE`
* `FLOAT`

### `datetime.datetime`

For `datetime` we will check `pandas.isnull` if it's available and return `NULL` if it returns true.
We also support `+inf` and `-inf` conversions.

If the `datetime` has tzinfo, we will use `TIMESTAMPTZ`, otherwise it becomes `TIMESTAMP`.

### `datetime.time`

If the `time` has tzinfo, we will use `TIMETZ`, otherwise it becomes `TIME`.

### `datetime.date`

> We support `+inf` and `-inf` conversions.

Author: This might be a little confusing, and needs some context
What I mean by this is that we check for these values and treat them as `infinity` and `-infinity`:

```c++
bool PyDate::IsPositiveInfinity() const {
	return year == 9999 && month == 12 && day == 31;
}

bool PyDate::IsNegativeInfinity() const {
	return year == 1 && month == 1 && day == 1;
}
```

Author: These are taken from:
```py
>>> import datetime
>>> datetime.date.max 
datetime.date(9999, 12, 31)
>>> datetime.date.min 
datetime.date(1, 1, 1)
```

Author: `datetime` has the same treatment, but the values are of course slightly different:
```py
datetime.datetime(1, 1, 1, 0, 0)
>>> datetime.datetime.max
datetime.datetime(9999, 12, 31, 23, 59, 59, 999999)
```

# example setting the sample size to 100000
duckdb.default_connection.execute("SET GLOBAL pandas_analyze_sample=100000")
```

### Object Conversion

> This is a mapping of python object type to DuckDB logical type:

Reviewer: ```suggestion
This is a mapping of Pyhon object types to DuckDB logical types:
```

`None` -> `NULL`  
`bool` -> `BOOLEAN`  

#### `int`
Since integers can be of arbitrary size in Python, there is not a one-to-one conversion possible for ints.  
Intead we perform these casts in order until one succeeds:  

> - `BIGINT`

Reviewer: is hugeint going to be tried once its in?

Author: Hugeint has been in for a while, it's not tried in this cast yet though.
When this was implemented it seemed too complex

No one has complained about it yet, when they do I'll have another look

- `INTEGER`
- `UBIGINT`
- `UINTEGER`
- `DOUBLE`

> When using the DuckDB Value class it's possible to set a target type, which will influence the conversion.  

Reviewer: ```suggestion
When using the DuckDB Value class, it's possible to set a target type, which will influence the conversion.  
```

#### `float`
These casts are tried in order:  
- `DOUBLE`  
- `FLOAT`  

`decimal.Decimal` -> `DECIMAL` / `DOUBLE`  
`uuid.UUID` -> `UUID`

#### `datetime.datetime`

> For datetimes we will check `pandas.isnat` if it's available and return `NULL` if it returns true.  

Reviewer 2: ```suggestion
For datetimes we will check `numpy.isnat` if it's available and return `NULL` if it returns true.  
```

> For datetimes we will check `pandas.isnat` if it's available and return `NULL` if it returns true.  

Reviewer: ```suggestion
For datetimes we will check `pandas.isnull` if it's available and return `NULL` if it returns true.
```

We also support `+inf` and `-inf` conversions.  

> If it has tzinfo we will use `TIMESTAMPTZ`, otherwise it becomes `TIMESTAMP`.  

Reviewer: ```suggestion
If the `datetime` has tzinfo, we will use `TIMESTAMPTZ`, otherwise it becomes `TIMESTAMP`.  
```

#### `datetime.time`

> If it has tzinfo we will use `TIMETZ`, otherwise it becomes `TIME`.  

Reviewer: ```suggestion
If the `time` has tzinfo, we will use `TIMETZ`, otherwise it becomes `TIME`.  
```

#### `datetime.date`
We support `+inf` and `-inf` conversions.  
It maps to the `DATE` type.  

#### `datetime.timedelta`
It maps to the `INTERVAL` type.  

> 

Reviewer: ```suggestion
It Python `timedelta` type maps to the `INTERVAL` DuckDB type.  
```

`str` -> `VARCHAR`  
`bytearray` -> `BLOB`  
`memoryview` -> `BLOB`  

#### `bytes`

> It converts to `BLOB` by default, when it's used to construct a Value object of type `BITSTRING` it maps to `BITSTRING` instead.  

Reviewer: ```suggestion
It converts to `BLOB` by default, when it's used to construct a Value object of type `BITSTRING`, it maps to `BITSTRING` instead.  
```

#### `list`

> It becomes the "biggest" type of its children, for example:  

Reviewer: ```suggestion
It becomes the "most permissive" type of its children, for example:  
```

```py
my_list_value = [
	12345,
	'test'
]
```
Will become `VARCHAR[]` because 12345 can convert to `VARCHAR` but `test` can not convert to `INTEGER`.  

#### `dict`

> This object can convert to either `STRUCT(...)` or `MAP(..., ...)` depending on its structure.  

Reviewer: ```suggestion
The `dict` object can convert to either `STRUCT(...)` or `MAP(..., ...)` depending on its structure.  
```

If the dict has a structure similar to:  
```py
my_map_dict = {
	'keys': [
		1, 2, 3
	],
	'values': [
		'one', 'two', 'three'
	]
}
```

> Then we'll convert it as a `MAP` of key-value pairs of the two lists zipped together.  

Reviewer: ```suggestion
Then we'll convert it to a `MAP` of key-value pairs of the two lists zipped together.  
```

> NOTE: the name of the fields matters and the two lists need to have the same size.  

Reviewer: ```suggestion
> The name of the fields matters and the two lists need to have the same size.  
```

Otherwise we'll try to convert it as a `STRUCT`  

> Where every key of the dictionary is converted to `str` to form the field names of the STRUCT.  

Reviewer: ```suggestion
Where every key of the dictionary is converted to `str` to form the field names of the `STRUCT`.
```

The fields are populated by converting every value of the dictionary.  

#### `tuple`
It converts to `LIST` by default, when it's used to construct a Value object of type `STRUCT` it maps to `STRUCT` instead.  

#### `numpy.ndarray` | `numpy.datetime64`

> It's converted by calling `tolist()` and converting the result of that.  

Reviewer: ```suggestion
`ndarray` and `datetime64` are converted by calling `tolist()` and converting the result of that.  
```
